lib/common/xml.c

```diff
@@ -460,29 +460,47 @@ pcmk__xe_first_child(const xmlNode *parent, const char *node_name,
     return NULL;
 }
 
-void
-copy_in_properties(xmlNode *target, const xmlNode *src)
+/*!
+ * \internal
+ * \brief Copy XML attributes, expanding \c ++ and \c += and checking ACLs
+ *
+ * This is similar to \c xmlCopyPropList(), with at least two notable
+ * differences:
+ * * \c ++ and \c += are expanded where appropriate. See \c expand_plus_plus()
+ *   for details.
+ * * This function returns immediately if ACLs prevent any attribute from being
+ *   copied to \p target.
+ *
+ * \param[in,out] target  XML element to receive copied attributes from \p src
+ * \param[in]     src     XML element whose attributes to copy to \p target
+ *
+ * \return Standard Pacemaker return code
+ */
+int
+pcmk__xe_copy_attrs(xmlNode *target, const xmlNode *src)
 {
-    if (src == NULL) {
-        crm_warn("No node to copy properties from");
+    CRM_CHECK((src != NULL) && (target != NULL), return EINVAL);
 
-    } else if (target == NULL) {
-        crm_err("No node to copy properties into");
+    for (xmlAttr *attr = pcmk__xe_first_attr(src); attr != NULL;
+         attr = attr->next) {
 
-    } else {
-        for (xmlAttrPtr a = pcmk__xe_first_attr(src); a != NULL; a = a->next) {
-            const char *p_name = (const char *) a->name;
-            const char *p_value = pcmk__xml_attr_value(a);
+        const char *name = (const char *) attr->name;
+        const char *value = pcmk__xml_attr_value(attr);
 
-            expand_plus_plus(target, p_name, p_value);
-            if (xml_acl_denied(target)) {
-                crm_trace("Cannot copy %s=%s to %s", p_name, p_value, target->name);
-                return;
-            }
+        expand_plus_plus(target, name, value);
+        if (xml_acl_denied(target)) {
```

Comment:
@kgaillot I think we can drop this check, as long as we give `copy_in_properties()` its own body so that the behavior is preserved there. `crm_xml_add()` checks ACLs. Between that and whatever final checking happens in `cib_perform_op()`, we should be fine. Let me know if you disagree.

Reply:
That would make it easier to reason about `pcmk__xe_copy_attrs()`, and would leave it with only one "extra" side effect: `++` expansion

Reply:
It seems OK as long as `update_cib_object()` still checks `xml_acl_denied()` after calling this.

Reply:
Thanks. The way I look at it, in places where we have
```
crm_xml_add(xml, name1, value1);
crm_xml_add(xml, name2, value2);
```

We don't call `xml_acl_denied()` in between those calls to determine whether to proceed. So it doesn't make much sense to do it when looping over a bunch of attrs, except for a tiny efficiency boost.

Reply:
`update_cib_object()` shouldn't need to check `xml_acl_denied()` either. Its only caller is `add_cib_object()`, whose only caller is `cib_process_create()`, which always goes through `cib_perform_op()`.

Reply:
Nevermind, I see what you mean, I forgot that it used to check `xml_acl_denied()` for setting its return value.

Reply:
> Thanks. The way I look at it, in places where we have
> 
> ```
> crm_xml_add(xml, name1, value1);
> crm_xml_add(xml, name2, value2);
> ```
> 
> We don't call `xml_acl_denied()` in between those calls to determine whether to proceed. So it doesn't make much sense to do it when looping over a bunch of attrs, except for a tiny efficiency boost if there is a denial.

ACLs are element-level, so the first attribute checked will be definitive.

Reply:
Right. We *can* check and short-circuit on ACL denial. What I'm saying is that it seems unnecessary, and that we *don't* do that in other contexts where we set multiple attributes for an element.